llvm/lib/SYCLLowerIR/ComputeModuleProperties.cpp

```diff
@@ -0,0 +1,380 @@
+      }
+
+      uint32_t WorkGroupNumDim = getKernelWorkGroupNumDim(Func);
+      if (WorkGroupNumDim) {
+        MetadataNames.push_back(Func.getName().str() + "@work_group_num_dim");
+        PropSet.add(PropSetRegTy::SYCL_PROGRAM_METADATA, MetadataNames.back(),
+                    WorkGroupNumDim);
+      }
+    }
+
+    // Add global_id_mapping information with mapping between device-global
+    // unique identifiers and the variable's name in the IR.
+    for (auto &GV : M.globals()) {
+      if (!isDeviceGlobalVariable(GV))
+        continue;
+
+      StringRef GlobalID = getGlobalVariableUniqueId(GV);
+      MetadataNames.push_back(GlobalID.str() + "@global_id_mapping");
+      PropSet.add(PropSetRegTy::SYCL_PROGRAM_METADATA, MetadataNames.back(),
+                  GV.getName());
+    }
+  }
+  bool SeenESIMDFunction = false;
+  bool SeenSYCLFunction = false;
+  for (const auto &F : M) {
+    if (llvm::module_split::isESIMDFunction(F))
+      SeenESIMDFunction = true;
+    else if (utils::isSYCLExternalFunction(&F) &&
+             !F.getName().starts_with("__itt"))
+      SeenSYCLFunction = true;
+  }
+  if (SeenESIMDFunction && !SeenSYCLFunction)
+    PropSet.add(PropSetRegTy::SYCL_MISC_PROP, "isEsimdImage", true);
+  {
+    StringRef RegAllocModeAttr = "sycl-register-alloc-mode";
+    uint32_t RegAllocModeVal;
+
+    bool HasRegAllocMode = llvm::any_of(EntryPoints, [&](const Function *F) {
+      if (!F->hasFnAttribute(RegAllocModeAttr))
+        return false;
+      const auto &Attr = F->getFnAttribute(RegAllocModeAttr);
+      RegAllocModeVal = getAttributeAsInteger<uint32_t>(Attr);
+      return true;
+    });
+    if (HasRegAllocMode) {
+      PropSet.add(PropSetRegTy::SYCL_MISC_PROP, RegAllocModeAttr,
+                  RegAllocModeVal);
+    }
+  }
+
+  {
+    StringRef GRFSizeAttr = "sycl-grf-size";
+    uint32_t GRFSizeVal;
+
+    bool HasGRFSize = llvm::any_of(EntryPoints, [&](const Function *F) {
+      if (!F->hasFnAttribute(GRFSizeAttr))
+        return false;
+      const auto &Attr = F->getFnAttribute(GRFSizeAttr);
+      GRFSizeVal = getAttributeAsInteger<uint32_t>(Attr);
+      return true;
+    });
+    if (HasGRFSize) {
+      PropSet.add(PropSetRegTy::SYCL_MISC_PROP, GRFSizeAttr, GRFSizeVal);
+    }
+  }
+
+  // FIXME: Remove 'if' below when possible
+  // GPU backend has a problem with accepting optimization level options in form
+  // described by Level Zero specification (-ze-opt-level=1) when 'invoke_simd'
+  // functionality is involved. JIT compilation results in the following error:
+  //     error: VLD: Failed to compile SPIR-V with following error:
+  //     invalid api option: -ze-opt-level=O1
+  //     -11 (PI_ERROR_BUILD_PROGRAM_FAILURE)
+  // 'if' below essentially preserves the behavior (presumably mistakenly)
+  // implemented in intel/llvm#8763: ignore 'optLevel' property for images which
+  // were produced my merge after ESIMD split
+  if (!SeenESIMDFunction || !SeenSYCLFunction) {
```

Comment:
This logic is new. Previously we used `MD.getEntryPointGroup().Props.HasESIMD !=
      module_split::SyclEsimdSplitStatus::SYCL_AND_ESIMD`